### PR TITLE
Implement Features for Generating and Rendering Mock Data in Browser

### PR DESCRIPTION
**Pull request by adaorachi**
### Description:
This pull request introduces a significant enhancement to the project's UI and API functionality as outlined in the [project requirements](https://docs.google.com/document/d/1KOpEpYGnNsfsAi03Q6IHWrzHwnKp10PN/edit?usp=sharing&ouid=110152251024460509561&rtpof=true&sd=true), enabling the rendering of data points in the browser and facilitating interactive data display through point-clicking functionality. Additionally, it establishes a mock API to efficiently transmit the extensive dataset to the frontend.

### Features Implemented
- [UI] Generated mock data and render 1M points in a browser, utilizing the regl-scatterplot library (referenced via https://github.com/flekschas/regl-scatterplot). Given the challenge of finding suitable mock data, random number generation was employed to fulfill this requirement and subsequent ones.

- [UI] Implemented functionality to enable users to click on individual points within the rendered data visualization to access associated data, (instead of image, text, audio, video, I used the random points data).

- [Stretch] Implemented a stretch goal feature wherein a preview of the associated data is displayed when users zoom in sufficiently on the data visualization.

- [API]  Developed a mock API to facilitate the transmission of 1M points from the backend to the frontend.

### Challenges:

One of the primary challenges encountered during the implementation process was the difficulty in finding suitable mock data for the project requirements. As a workaround, random number generation was utilized to simulate the required data points.

### Test Plan:

- Verify that mock data is successfully generated and rendered in the browser interface.
- Confirm that clicking on individual data points displays associated image, text, audio, or video data.
- Test stretch goal feature to ensure that previews of data details are shown when zoomed in sufficiently.
- Validate the functionality of the mock API in transmitting 1 million data points to the frontend.

### Screenshots/GIFs:

https://github.com/facebookresearch/MultiModalExplorer/assets/50773261/de6466f5-4ae0-4ee8-8c48-f5e48111bc0d

### CODE REVIEW

File: frontend/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# React + TypeScript + Vite
```
Comment (Mortimerp9) [inline frontend/README.md]: you might want to replace this template README with some basic instructions on how to run/build the frontend

File: frontend/public/vite.svg
Comment (Mortimerp9) [inline frontend/public/vite.svg:1]: we probably don't want to include this

File: frontend/src/actions/api/embedding.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+};
+
+export const getEmbeddingPointDetails = async (
+  point: number[]
+): Promise<{ id: number }> => {
+  return new Promise((resolve, reject) => {
+    try {
+      setTimeout(() => {
+        resolve({ id: point[0] });
+      }, 2000);
```
Comment (Mortimerp9) [inline frontend/src/actions/api/embedding.ts:30]: can you add TODO when it's mocked

File: frontend/README.md (hunk 1 above)
Comment (adaorachi) [reply frontend/README.md]: Yes, that was an honest oversight. I'll add the necessary instructions right away.

File: frontend/public/vite.svg
Comment (adaorachi) [reply frontend/public/vite.svg:1]: Is it okay I replace it with a Meta favicon or svg?

File: frontend/src/actions/api/embedding.ts (hunk 2 above)
Comment (adaorachi) [reply frontend/src/actions/api/embedding.ts:30]: Okay, I'll do that.

File: frontend/README.md (hunk 1 above)
Comment (adaorachi) [reply frontend/README.md]: I've made a new push and the README has been updated.

File: frontend/public/vite.svg
Comment (adaorachi) [reply frontend/public/vite.svg:1]: The vite.svg has been removed.

File: frontend/src/actions/api/embedding.ts (hunk 2 above)
Comment (adaorachi) [reply frontend/src/actions/api/embedding.ts:30]: This has been added.

File: frontend/public/vite.svg
Comment (Mortimerp9) [reply frontend/public/vite.svg:1]: I am not sure that we allowed. It's ok if we don't have a custom favicon for now.

**Comment by Mortimerp9** [comment]
I know that a lot of these files come straight from the template. for files to be in this repo we need to keep a copyright header on them:
``` 
/**
 * (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.
 */
```

**Comment by adaorachi** [comment]
> I know that a lot of these files come straight from the template. for files to be in this repo we need to keep a copyright header on them:
> 
> ```
> /**
>  * (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.
>  */
> ```

So, just to be clear, what I need to do is add this header to all generated files, right?

**Comment by Mortimerp9** [comment]
yes, all files that contain code have to have a license header.

File: frontend/src/App.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  );
+
+  const [embeddingPoint, setEmbeddingPoint] = useState<null | number[]>(null);
+
+  const [embeddingDetails, setEmbeddingDetails] = useState<null | {
+    id: number;
+  }>(null);
+
+  const [loadingEmbeddingDetails, setLoadingEmbeddingDetails] =
+    useState<boolean>(false);
```
Comment (somya) [inline frontend/src/App.tsx]: I recommend adding a comment for each of these consts to make it clearer what they are for. e.g what's an embedding point vs  embeddingDetails? Why is embeddingPoint an array?

File: frontend/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  useEffect(() => {
+    const embeddings = async () => {
+      const resp = await getEmbeddingPoints();
```
Comment (somya) [inline frontend/src/App.tsx]: Nice, I like how you've broken out the actions separately.

File: frontend/src/App.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      setEmbeddings(resp);
+    };
+    embeddings();
```
Comment (somya) [inline frontend/src/App.tsx]:
```suggestion
    fetchEmbeddings();
```
embeddings alone is very ambigious.

File: frontend/src/App.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    setLoadingEmbeddingDetails(true);
+    setEmbeddingPoint(point);
+
+    const details = await getEmbeddingPointDetails(point);
+
+    setEmbeddingDetails(details);
+    setLoadingEmbeddingDetails(false);
+  };
+
+  const handlePointUnclicked = () => {
```
Comment (somya) [inline frontend/src/App.tsx]:
```suggestion
  const handleEmdeddingUnselect = () => {
```
does "handlePointUnselected" better convey this function? "Unclicked" seems a little odd. Also I'd prefer "embedding" over "point", but open to other ideas.

File: frontend/src/App.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  RenderEmbeddings,
+  Navbar,
+  EmbeddingDetails,
+  Footer,
+  Loader,
+} from "./components";
+import { getEmbeddingPoints, getEmbeddingPointDetails } from "./actions";
+import "./index.css";
+
+function App() {
```
Comment (somya) [inline frontend/src/App.tsx:12]: In a future diff I would suggest refactoring this into it's own component. So app can be used for higher level features such as router, header, footer etc.

File: frontend/src/actions/api/embedding.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+const BASE_URL = "http://127.0.0.1:8000/api";
+
+export const getEmbeddingPoints = async (): Promise<
+  Array<[number, number, number]>
+> => {
+  try {
+    const response = await axios.get(`${BASE_URL}/embeddings`);
+    return response?.data?.data || [];
+  } catch (error) {
+    return [];
```
Comment (somya) [inline frontend/src/actions/api/embedding.ts]: Let's display an error to the user. A simple approach is create a global error state variable and when set it displays a popup of some sort to report the error.

Comment (somya) [reply frontend/src/actions/api/embedding.ts]: Here you can just throw the error: `throw new Error(error)`

File: frontend/src/actions/api/embedding.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  // This is currently a mock to simulate behavior for testing purposes.
+  // TODO: Replace this mock with the real implementation once it's available.
+
+  return new Promise((resolve, reject) => {
+    try {
+      setTimeout(() => {
+        resolve({ id: point[0] });
+      }, 2000);
+    } catch (error) {
+      reject(error);
```
Comment (somya) [inline frontend/src/actions/api/embedding.ts:32]: oh ok reject is fine too.

File: frontend/src/actions/index.ts [hunk 10]
```diff
@@ -0,0 +1 @@
+export { getEmbeddingPoints, getEmbeddingPointDetails } from "./api/embedding";
```
Comment (somya) [inline frontend/src/actions/index.ts]: not sure this is needed, can't you just get the functions for the embeddings.ts  file?

File: frontend/src/components/RenderEmbeddings.tsx [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+          <canvas
+            ref={textOverlayRef}
+            className="!w-full !h-full absolute top-0 right-0 pointer-events-none"
+          />
+        </div>
+      ) : (
+        <div className="w-screen h-screen flex justify-center items-center">
+          <Loader />
+        </div>
+      )}
```
Comment (somya) [inline frontend/src/components/RenderEmbeddings.tsx]: I thought you had a loader in App.tsx, Do you need another one here?

File: frontend/src/components/index.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+export { default as Navbar } from "./Navbar";
+export { default as Loader } from "./Loader";
+
+export { default as RenderEmbeddings } from "./RenderEmbeddings";
+export { default as EmbeddingDetails } from "./EmbeddingDetails";
+export { default as Footer } from "./Footer";
```
Comment (somya) [inline frontend/src/components/index.ts]: I don't think you need index files. But LMK if I'm missing something.

**Review by somya — APPROVED** [review]
Hi @adaorachi Overall this looks great! 

I mainly have minor feedback in terms of naming, adding comments when needed. Feel free to merge when ready.

**Comment by adaorachi** [comment]
> Hi @adaorachi Overall this looks great!
> 
> I mainly have minor feedback in terms of naming, adding comments when needed. Feel free to merge when ready.

Thank you @somya for your extensive review.
I'll begin working on implementing the recommended changes.

File: frontend/src/components/index.ts (hunk 12 above)
Comment (adaorachi) [reply frontend/src/components/index.ts]: Placing the index inside a folder helps to improve my code organization, simplifies imports and provides a entry point for the folder's functionality. However, I'm open to using alternative approaches as well.

**Review by s0mya — APPROVED** [review]
LGTM!

**Pull request merged**